### PR TITLE
Update webpack 5

### PR DESCRIPTION
**Pull request by danielegiorgis**
WIP for webpack 5, so you can view and evaluate.
So far the changes are not many, but as regards the cli some flags have changed and I have not yet been able to identify the new correspondents.

### CODE REVIEW

**Comment by pioneerskies** [comment]
First thing first: ❤️ 

How do you associate, if you do, this one with #256 ?
Are you planning to insert this one into 6.0 release or do you have other plans?

Thanks for your effort :) I'll read and eventually comment the code asap. in the meantime I'll take note of the webpack 5 release announcement https://webpack.js.org/blog/2020-10-10-webpack-5-release/ in order to better align myself with the news

**Comment by danielegiorgis** [comment]
Ok, I removed Imagemin in favor of the new imagemin webpack (image-minimizer-webpack-plugin)
https://github.com/webpack-contrib/image-minimizer-webpack-plugin
As I wrote in the commit, the change was necessary because the webpack warning "BREAKING CHANGE: No more changes should happen to Compilation.assets after sealing the Compilation. Imagemin"
In addition Imagemin is deprecated and the author himself refers to the new plugin.

In addition it was necessary to enter publicPath: "" for the options of MiniCssExtractPlugin due to the error during the builds
"ERROR in ./src/stylesheets/screen.scss
Module build failed (from ./node_modules/mini-css-extract-plugin/dist/loader.js):
Error: Automatic publicPath is not supported in this browser "

I did some build tests and even with subfolders the .css is correctly included and the resultant correctly inserted in dist.
However, I ask you to consider if what you have done is okay <3

**Comment by danielegiorgis** [comment]
> First thing first: ❤️
> 
> How do you associate, if you do, this one with #256 ?
> Are you planning to insert this one into 5.0 release or do you have other plans?
> 
> Thanks for your effort :) I'll read and eventually comment the code asap. in the meantime I'll take note of the webpack 5 release announcement https://webpack.js.org/blog/2020-10-10-webpack-5-release/ in order to better align myself with the news

As I said in the comment https://github.com/welaika/wordless/pull/256#issuecomment-748512707 I've merged that branch in this one

File: wordless/theme_builder/vanilla_theme/webpack.env.js [hunk 1]
```diff
@@ -6,7 +6,7 @@ devOptions = {
     {
       loader: MiniCssExtractPlugin.loader,
       options: {
-        sourceMap: true
```
Comment (pioneerskies) [inline wordless/theme_builder/vanilla_theme/webpack.env.js:9]: why did you remove the `sourceMap` key? is it moved somewhere else? given this update is neither necessary to put the configuration inside `.env.js` part of the configuration, since they are equal for both envoronment.
But I think that switching on/off by environment the generation of source maps was intended as a feature.

Comment (danielegiorgis) [reply wordless/theme_builder/vanilla_theme/webpack.env.js:9]: Because,

ERROR in ./src/stylesheets/screen.scss
Module build failed (from ./node_modules/mini-css-extract-plugin/dist/loader.js):
ValidationError: Invalid options object. Mini CSS Extract Plugin Loader has been initialized using an options object that does not match the API schema.
 - options has an unknown property 'sourceMap'. These properties are valid:
 
 and I didn't found a solution yet

Comment (pioneerskies) [reply wordless/theme_builder/vanilla_theme/webpack.env.js:9]: As discussed elsewhere, sourcemap production can be controlled directly through `devtools` key, so it's just ok to have removed this config line

File: wordless/theme_builder/vanilla_theme/webpack.config.js [hunk 2]
```diff
@@ -22,7 +24,8 @@ module.exports = (env) => {
 
     output: {
       filename: "[name].js",
-      path: javascriptsDstPath
+      path: javascriptsDstPath,
+      publicPath: path.join('/wp-content/themes', path.basename(__dirname), 'dist/')
```
Comment (pioneerskies) [inline wordless/theme_builder/vanilla_theme/webpack.config.js:28]: This is equal to simply use `dstDir` variable

File: wordless/theme_builder/vanilla_theme/webpack.config.js [hunk 3]
```diff
@@ -61,7 +64,11 @@ module.exports = (env) => {
           test: /\.(jpe?g|png|gif|svg)$/i,
           loader: 'file-loader',
           options: {
-            name: '[path][name].[ext]',
+            name (resourcePath) {
+              resourcePath = resourcePath.replace(path.join(srcDir, `/${imageFolder}/`), '');
+              return resourcePath;
+            },
+            outputPath: imageFolder
```
Comment (pioneerskies) [inline wordless/theme_builder/vanilla_theme/webpack.config.js:71]:
```suggestion
            name: '[name].[ext]',
            context: path.join(srcDir),
            outputPath: (_url, resourcePath, context) => {
              return resourcePath.replace(context, '')
            }
```

this suggested code is just equivalent to the previous one, but the function to manage the custom logic is moved to the `outputPath` key, which is more striclty responsible to choose the path; while the `name` key is really and only just responsible of the file naming.
I've also used a custom `context` just in order to diminsh the complexity inside the function.

It's really just a matter of readibility, not functionality. So feel free to choose your favourite approach.

File: wordless/theme_builder/vanilla_theme/webpack.config.js [hunk 4]
```diff
@@ -7,6 +7,8 @@ const BrowserSyncPlugin = require('browser-sync-webpack-plugin');
 const MiniCssExtractPlugin = require('mini-css-extract-plugin');
 const ImageminWebpack = require('image-minimizer-webpack-plugin');
 const CopyWebpackPlugin = require('copy-webpack-plugin');
+
+const imageFolder = 'images'
```
Comment (pioneerskies) [inline wordless/theme_builder/vanilla_theme/webpack.config.js:11]: I'd prefere to name it `imageFolderName` in order to distinguish it from the various paths already set up.

Alternatively having a `imagesDstPath` and a `imagesSrcPath` set up to the complete path could be another valid approach and would follow the previously adopted practice. But this could be just overwhelming.

File: wordless/theme_builder/vanilla_theme/webpack.env.js [hunk 5]
```diff
@@ -5,9 +5,7 @@ devOptions = {
   styleLoaders: [
     {
       loader: MiniCssExtractPlugin.loader,
-      options: {
-        publicPath: '',
-      }
+      options: {}
```
Comment (pioneerskies) [inline wordless/theme_builder/vanilla_theme/webpack.env.js]: Probably I'd opt to remove the empty options. Here and in line 32.

Comment (danielegiorgis) [reply wordless/theme_builder/vanilla_theme/webpack.env.js]: I've done a refactor in https://github.com/welaika/wordless/pull/261/commits/6b95f16c66fc6541d5a955a20bcedf7abee63dc0

File: wordless/theme_builder/vanilla_theme/webpack.config.js [hunk 6]
```diff
@@ -111,17 +103,37 @@ module.exports = (env) => {
...
-          from: path.join(srcDir, '/images'),
-          to: path.join(dstDir, '/images', '[path][name].[ext]'),
-          toType: 'template'
-        }
-      ])
+      new CopyWebpackPlugin({
+        patterns: [
+          {
+            from: path.join(srcDir, `/${imageFolder}`),
+            to: path.join(dstDir, `/${imageFolder}`, '[path][name].[ext]'),
```
Comment (pioneerskies) [inline wordless/theme_builder/vanilla_theme/webpack.config.js:132]: I think that `[path][name].[ext]` is an error 'coz you'd need a `/` betweeb `path` and `name` placeholders. Isn't it correct?

Moreover: if we'd like to follow my suggestions about `file-loader` config, we should also update here to

```suggestions
            to: path.join(dstDir, `/${imageFolder}`, '[folder]', '[name].[ext]'),
```

But this depens on the choice about the other piece of code.

File: wordless/theme_builder/vanilla_theme/webpack.env.js (hunk 5 above)
Comment (pioneerskies) [reply wordless/theme_builder/vanilla_theme/webpack.env.js]: hell yeah

File: wordless/theme_builder/vanilla_theme/package.json [hunk 7]
```diff
@@ -1,19 +1,19 @@
...
   "license": "MIT",
   "private": true,
   "engines": {
-    "node": "12.13.1"
+    "node": "14.15.3"
   },
   "scripts": {
     "server": "npx nf start",
-    "build:dev": "webpack --debug --env.NODE_ENV=development",
-    "build:prod": "yarn sign-release && webpack -p --bail --env.NODE_ENV=production",
```
Comment (pioneerskies) [inline wordless/theme_builder/vanilla_theme/package.json:16]: Do we have an alternative to `--bail` in webpack5?

File: wordless/theme_builder/vanilla_theme/package.json [hunk 8]
```diff
@@ -1,19 +1,19 @@
...
   "author": "'weLaika' <info@welaika.com>",
   "license": "MIT",
   "private": true,
   "engines": {
-    "node": "12.13.1"
+    "node": "14.15.3"
   },
   "scripts": {
     "server": "npx nf start",
-    "build:dev": "webpack --debug --env.NODE_ENV=development",
```
Comment (pioneerskies) [inline wordless/theme_builder/vanilla_theme/package.json:15]: Do we have an alternative to `--debug` in webpack5?

File: wordless/theme_builder/vanilla_theme/Procfile [hunk 9]
```diff
@@ -1,3 +1,3 @@
 wp: wp server --host=0.0.0.0
-webpack: npx webpack --debug --watch --progress --color --env.NODE_ENV=development
```
Comment (pioneerskies) [inline wordless/theme_builder/vanilla_theme/Procfile:2]: Do we have an alternative to --debug in webpack5?

File: wordless/theme_builder/vanilla_theme/package.json (hunk 7 above)
Comment (pioneerskies) [inline wordless/theme_builder/vanilla_theme/package.json:16]: Do we have an alternative to `-p` in webpack5?

**Review by pioneerskies — COMMENTED** [review]
Dive a bit into the no more supported webpack cli's flags

File: wordless/theme_builder/vanilla_theme/Procfile (hunk 9 above)
Comment (danielegiorgis) [inline wordless/theme_builder/vanilla_theme/Procfile:2]: Need to understand if this can be a problem.

with webpack5 and the new webpack-cli we don't have some flags 
from the documentation @v.4

--debug --> Switch loaders to debug mode
--bail --> Abort the compilation on first error
--p --optimize-minimize --define process.env.NODE_ENV="production"

File: wordless/theme_builder/vanilla_theme/package.json (hunk 7 above)
Comment (pioneerskies) [reply wordless/theme_builder/vanilla_theme/package.json:16]: I've seen the new config. Weeeeee

**Review by pioneerskies — APPROVED** [review]
